### PR TITLE
added additional instance cache metric

### PR DESCRIPTION
**Pull request by troy4eg**
added metric to track the amount of memory that is used for the store instance to instance cache

### CODE REVIEW

File: runtime/interface.h [hunk 1]
```diff
@@ -243,6 +243,11 @@ inline void f$kphp_turn_on_host_tag_in_inner_statshouse_metrics_toggle() {
   StatsHouseManager::get().turn_on_host_tag_toggle();
 }
 
+template <typename F>
+inline void f$kphp_extended_instance_cache_metrics_init(F &&callback) {
+  StatsHouseManager::get().set_normalization_function(normalization_function{std::forward<F>(callback)});
```
Comment (DrDet) [inline runtime/interface.h:248]: we need to add critical section here to guard possible heap usage in `std::function` constructor

File: server/statshouse/statshouse-manager.h [hunk 2]
```diff
@@ -14,6 +14,8 @@
 #include "server/workers-control.h"
 #include "server/workers-stats.h"
 
+using normalization_function = std::function<string(const string &)>;
```
Comment (DrDet) [inline server/statshouse/statshouse-manager.h:17]: let's `#include <functional>` above

File: server/statshouse/statshouse-manager.h [hunk 3]
```diff
@@ -42,6 +44,14 @@ class StatsHouseManager : vk::not_copyable {
     need_write_enable_tag_host = true;
   }
 
+  void set_normalization_function(normalization_function &&_function) {
+    this->key_normalization_function = _function;
```
Comment (DrDet) [inline server/statshouse/statshouse-manager.h:48]:
```suggestion
    this->key_normalization_function = std::move(_function);
```

File: server/statshouse/statshouse-manager.h [hunk 4]
```diff
@@ -61,9 +71,15 @@ class StatsHouseManager : vk::not_copyable {
...
 
+  /**
+   * before calling the method, be sure to is_extended_instance_cache_stats_enabled() is true
+   */
+  void add_extended_instance_cache_stats(const std::string_view &type, const std::string_view &status, const string &key, uint64_t size = 0);
+
 private:
   StatsHouseClient client;
   bool need_write_enable_tag_host = false;
+  normalization_function key_normalization_function = nullptr;
```
Comment (DrDet) [inline server/statshouse/statshouse-manager.h:82]: let's add 'instance_cache' somewhere to the name

File: server/statshouse/statshouse-manager.h [hunk 5]
```diff
@@ -61,9 +71,15 @@ class StatsHouseManager : vk::not_copyable {
   void add_common_master_stats(const workers_stats_t &workers_stats, const memory_resource::MemoryStats &memory_stats, double cpu_s_usage, double cpu_u_usage,
                                long long int instance_cache_memory_swaps_ok, long long int instance_cache_memory_swaps_fail);
 
+  /**
+   * before calling the method, be sure to is_extended_instance_cache_stats_enabled() is true
+   */
+  void add_extended_instance_cache_stats(const std::string_view &type, const std::string_view &status, const string &key, uint64_t size = 0);
```
Comment (DrDet) [inline server/statshouse/statshouse-manager.h:77]: It's very strange to pass `std::string_view` by reference. It's a view, so it's already a reference

File: runtime/instance-cache.h [hunk 6]
```diff
@@ -89,10 +113,13 @@ ClassInstanceType f$instance_cache_fetch(const string &class_name, const string
     if (auto wrapper = dynamic_cast<const InstanceCopyistImpl<ClassInstanceType> *>(base_wrapper)) {
       auto result = wrapper->get_instance();
       php_assert(!result.is_null());
+      send_extended_instance_cache_stats_if_enabled("fetch", InstanceCacheStoreStatus::success, key, result);
```
Comment (DrDet) [inline runtime/instance-cache.h:116]: it's strange to use `InstanceCacheStoreStatus` enum as fetch status

File: runtime/instance-cache.h [hunk 7]
```diff
@@ -89,10 +113,13 @@ ClassInstanceType f$instance_cache_fetch(const string &class_name, const string
...
       return result;
     } else {
       php_warning("Trying to fetch incompatible instance class: expect '%s', got '%s'",
                   class_name.c_str(), base_wrapper->get_class());
+      ClassInstanceType empty = {};
+      send_extended_instance_cache_stats_if_enabled("fetch", InstanceCacheStoreStatus::failed, key, empty);
```
Comment (DrDet) [inline runtime/instance-cache.h:122]:
```suggestion
      send_extended_instance_cache_stats_if_enabled("fetch", InstanceCacheStoreStatus::failed, key, ClassInstanceType{});
```

File: server/statshouse/statshouse-manager.h (hunk 5 above)
Comment (troy4eg) [reply server/statshouse/statshouse-manager.h:77]: Ops, initially it was just a const string&

File: server/statshouse/statshouse-manager.h [hunk 8]
```diff
@@ -66,9 +77,15 @@ class StatsHouseManager : vk::not_copyable {
    */
   void add_init_master_stats(uint64_t total_init_ns, uint64_t confdata_init_ns);
 
+  /**
+   * before calling the method, be sure to is_extended_instance_cache_stats_enabled() is true
+   */
+  void add_extended_instance_cache_stats(const std::string_view type, const std::string_view status, const string &key, uint64_t size = 0);
```
Comment (DrDet) [inline server/statshouse/statshouse-manager.h:83]: `std::string_view` is always const
```suggestion
  void add_extended_instance_cache_stats(std::string_view type, std::string_view status, const string &key, uint64_t size = 0);
```

**Review by DrDet — APPROVED** [review]

**Review by DrDet — APPROVED** [review]

**Review by DrDet — APPROVED** [review]

**Pull request merged**